### ISSUE TITLE
ensure #[pallet::constant] is actually constant

### ISSUE THREAD

**Issue by xlc** [paritytech/substrate#11619]
It is possible to annotate a field with `#[pallet::constant]` which is actually not a constant and causing unexpected behaviour.

We should have a way to assert the pallet constant are actually constant. Not exactly sure how it can be implemented but there must be some way.

Related: https://github.com/paritytech/polkadot/issues/4015 https://github.com/AcalaNetwork/Acala/issues/2184

**Comment by nazar-pc** [comment]
We actually (ab)use the fact that it isn't actually a constant (at least in a sense that we can have different values without updating runtime) for implementing dynamic cost of storage (our `LengthToFee` is not constant, it is derived from consensus and changes every block).

Otherwise we'll need to refactor and make `LengthToFee` non-constant and that has significant implications.

**Comment by xlc** [comment]
In that case it should not be annotated with `#[pallet::constant]` and provide a different way to expose the value.

**Comment by nazar-pc** [comment]
Well, Bastian suggested that it is fine as long as value stays the same for the duration of a particular block, so we went with that.
I did find it counter-intuitive as well.

**Comment by bkchr** [comment]
> Well, Bastian suggested that it is fine as long as value stays the same for the duration of a particular block, so we went with that. I did find it counter-intuitive as well.

Did I? :D I can not remember :P 

To the issue, I'm with you @xlc, but it will maybe also be hard to solve.Using the example of @nazar-pc, we may would need to introduce `LengthToFeeStatic` and a `LengthToFeeDynamic`. While one needs to be set and the other one could be `None` or something. However, I think there are more of these things hidden in the code base.

A trick to check if constants are really constant could be to `get` each constant and return an error if we accessed the storage while doing so.

**Comment by xlc** [comment]
I guess the short term solution will be have `#[pallet::constant]` generate an integrity test that calls get and ensure it doesn't access storages.

**Comment by nazar-pc** [comment]
But in our protocol it does for `LengthToFee` and will break our codebase,

**Comment by xlc** [comment]
It is not a constant so you should be fine

https://github.com/paritytech/substrate/blob/a0a40adbf57b47b445012c6cbca940c991c94197/frame/transaction-payment/src/lib.rs#L287-L288

**Comment by xlc** [comment]
I can totally see this breaks a lot of chains, so it could be an opt-in feature to start with.

**Comment by kianenigma** [comment]
`pallet::constant` in its current form should really be renamed `pallet::metadata` foremost, as it does not imply being constant and instead is about just being exposed in metadata.

**Comment by xlc** [comment]
Metadata should be constant within a version. Otherwise, it won't play nicely with features like https://github.com/paritytech/substrate/issues/10478

**Comment by bkchr** [comment]
> Metadata should be constant within a version. Otherwise, it won't play nicely with features like #10478

Yeah, I think to ensure this I would introduce some kind of "no state access" runtime api calls. Then we could move the `metadata` to use this. If there would be a state access, it would fail.
